### DISCUSSION TITLE
Cannot consume internal events published to amq.rabbitmq.events

### DISCUSSION THREAD

**Discussion by eliyosef578** [rabbitmq/rabbitmq-server#12053] [Other] [answered]
### Describe the bug

hey,
I am attempting to initialize RabbitMQ with a deployment that updates two components while RabbitMQ is running:
 the plugin list and the definition file.
 After the upgrade, the plugin named "rabbitmq-event-exchange" is not functioning correctly.
 Users who manually bind to the queue are not receiving events on the requested queue.
 If I do not enable the plugin during the upload process and instead enable it manually using the "rabbitmq-plugins enable" command, the events are delivered to the queue as expected.
 Before the upgrade, we were using version 3.8.19 and it was working properly.

### Reproduction steps

1.init "rabbitmq-event-exchang" with Enabled Plugins File
2.bind excahge.created to new queue from amq.rabbitmq.event
3.check if you are see message in the queue
...

### Expected behavior

the events are delivered to the queue as expected.

**Comment by michaelklishin** [comment]
@eliyosef578 that exchange type hasn't changed meaningfully in years.

I don't see how re-configured plugin list would be different from manually enabled plugins. There are so many systems that pre-configure plugins these days that we'd be flooded with questions like that but it is not the case.

My best guess is that your `enabled_plugins` file is incorrect and the node ignores it (or some of the plugins), therefore the exchange is not actually enabled. But you haven't provided any data for us to work with and we do not guess in this community.

Consider posting your `rabbitmq.conf` and `enabled_plugins` files, as well as the output of `rabbitmq-plugins list` in both cases.

**Comment by michaelklishin** [comment, marked as answer]
Given the following `enabled_plugins` file:

``` erl
[rabbitmq_management,rabbitmq_event_exchange].
```

and a 3.13.5 node started like so

``` shell
RABBITMQ_ENABLED_PLUGINS_FILE=./enabled_plugins rabbitmq-server
```

I see the following in the `rabbitmq-diagnostics status` output:

```
Plugins

Enabled plugin file: ./enabled_plugins
Enabled plugins:

 * rabbitmq_event_exchange
 * rabbitmq_management
 * rabbitmq_management_agent
 * rabbitmq_web_dispatch
 * amqp_client
 * cowboy
 * cowlib
 * oauth2_client
 * jose
```

Then I bind a quorum queue with routing key `#` (which, for a topic exchange like `amq.rabbitmq.event`, matches everything) to it, using Bunny for client and a Ruby REPL as an alternative to writing a script:

``` ruby
# open a connection and channel first, then…
qq = ch.quorum_queue("qq.1")
qq.bind("amq.rabbitmq.event", routing_key: "#").subscribe do |*args|
  puts(args.inspect)
end
```

and then declare and delete 100 queues:

``` ruby
100.times { |i| ch.queue("cq.transient.#{i}", durable: false).delete }
```

the consumer does print them out, and then prints a `nodes.node.stats` events every 5 seconds.

I can also see the following in `rabbitmq-diagnostics consume_event_stream` output:

```
{"event":"queue.created","type":"rabbit_classic_queue","user_who_performed_action":"guest","exclusive":false,"owner_pid":"","arguments":[],"auto_delete":false,"durable":false,"name":"cq.transient.88","vhost":"/","timestamp_in_ms":1724083324406}
{"event":"queue.deleted","user_who_performed_action":"guest","type":"rabbit_classic_queue","name":"cq.transient.88","vhost":"/","timestamp_in_ms":1724083324408}
{"event":"queue.created","type":"rabbit_classic_queue","user_who_performed_action":"guest","exclusive":false,"owner_pid":"","arguments":[],"auto_delete":false,"durable":false,"name":"cq.transient.89","vhost":"/","timestamp_in_ms":1724083324409}
{"event":"queue.deleted","user_who_performed_action":"guest","type":"rabbit_classic_queue","name":"cq.transient.89","vhost":"/","timestamp_in_ms":1724083324411}
{"event":"queue.created","type":"rabbit_classic_queue","user_who_performed_action":"guest","exclusive":false,"owner_pid":"","arguments":[],"auto_delete":false,"durable":false,"name":"cq.transient.90","vhost":"/","timestamp_in_ms":1724083324412}
{"event":"queue.deleted","user_who_performed_action":"guest","type":"rabbit_classic_queue","name":"cq.transient.90","vhost":"/","timestamp_in_ms":1724083324414}
{"event":"queue.created","type":"rabbit_classic_queue","user_who_performed_action":"guest","exclusive":false,"owner_pid":"","arguments":[],"auto_delete":false,"durable":false,"name":"cq.transient.91","vhost":"/","timestamp_in_ms":1724083324415}
{"event":"queue.deleted","user_who_performed_action":"guest","type":"rabbit_classic_queue","name":"cq.transient.91","vhost":"/","timestamp_in_ms":1724083324417}
{"event":"queue.created","type":"rabbit_classic_queue","user_who_performed_action":"guest","exclusive":false,"owner_pid":"","arguments":[],"auto_delete":false,"durable":false,"name":"cq.transient.92","vhost":"/","timestamp_in_ms":1724083324418}
{"event":"queue.deleted","user_who_performed_action":"guest","type":"rabbit_classic_queue","name":"cq.transient.92","vhost":"/","timestamp_in_ms":1724083324420}
{"event":"queue.created","type":"rabbit_classic_queue","user_who_performed_action":"guest","exclusive":false,"owner_pid":"","arguments":[],"auto_delete":false,"durable":false,"name":"cq.transient.93","vhost":"/","timestamp_in_ms":1724083324421}
{"event":"queue.deleted","user_who_performed_action":"guest","type":"rabbit_classic_queue","name":"cq.transient.93","vhost":"/","timestamp_in_ms":1724083324422}
{"event":"queue.created","type":"rabbit_classic_queue","user_who_performed_action":"guest","exclusive":false,"owner_pid":"","arguments":[],"auto_delete":false,"durable":false,"name":"cq.transient.94","vhost":"/","timestamp_in_ms":1724083324424}
{"event":"queue.deleted","user_who_performed_action":"guest","type":"rabbit_classic_queue","name":"cq.transient.94","vhost":"/","timestamp_in_ms":1724083324426}
{"event":"queue.created","type":"rabbit_classic_queue","user_who_performed_action":"guest","exclusive":false,"owner_pid":"","arguments":[],"auto_delete":false,"durable":false,"name":"cq.transient.95","vhost":"/","timestamp_in_ms":1724083324427}
{"event":"queue.deleted","user_who_performed_action":"guest","type":"rabbit_classic_queue","name":"cq.transient.95","vhost":"/","timestamp_in_ms":1724083324428}
{"event":"queue.created","type":"rabbit_classic_queue","user_who_performed_action":"guest","exclusive":false,"owner_pid":"","arguments":[],"auto_delete":false,"durable":false,"name":"cq.transient.96","vhost":"/","timestamp_in_ms":1724083324429}
{"event":"queue.deleted","user_who_performed_action":"guest","type":"rabbit_classic_queue","name":"cq.transient.96","vhost":"/","timestamp_in_ms":1724083324431}
{"event":"queue.created","type":"rabbit_classic_queue","user_who_performed_action":"guest","exclusive":false,"owner_pid":"","arguments":[],"auto_delete":false,"durable":false,"name":"cq.transient.97","vhost":"/","timestamp_in_ms":1724083324432}
{"event":"queue.deleted","user_who_performed_action":"guest","type":"rabbit_classic_queue","name":"cq.transient.97","vhost":"/","timestamp_in_ms":1724083324433}
{"event":"queue.created","type":"rabbit_classic_queue","user_who_performed_action":"guest","exclusive":false,"owner_pid":"","arguments":[],"auto_delete":false,"durable":false,"name":"cq.transient.98","vhost":"/","timestamp_in_ms":1724083324434}
{"event":"queue.deleted","user_who_performed_action":"guest","type":"rabbit_classic_queue","name":"cq.transient.98","vhost":"/","timestamp_in_ms":1724083324436}
{"event":"queue.created","type":"rabbit_classic_queue","user_who_performed_action":"guest","exclusive":false,"owner_pid":"","arguments":[],"auto_delete":false,"durable":false,"name":"cq.transient.99","vhost":"/","timestamp_in_ms":1724083324437}
{"event":"queue.deleted","user_who_performed_action":"guest","type":"rabbit_classic_queue","name":"cq.transient.99","vhost":"/","timestamp_in_ms":1724083324439}
{"event":"node.node.stats","send_bytes":97257,"recv_bytes":705,"sock_port":25672,"sock_addr":"127.0.0.1","peer_port":54109,"peer_addr":"127.0.0.1","route":"{rabbit@sunnyside,'rabbitmqcli-932-rabbit@sunnyside'}","timestamp_in_ms":1724083328304}
{"event":"node.node.stats","send_bytes":97519,"recv_bytes":705,"sock_port":25672,"sock_addr":"127.0.0.1","peer_port":54109,"peer_addr":"127.0.0.1","route":"{rabbit@sunnyside,'rabbitmqcli-932-rabbit@sunnyside'}","timestamp_in_ms":1724083333350}
{"event":"node.node.stats","send_bytes":97781,"recv_bytes":709,"sock_port":25672,"sock_addr":"127.0.0.1","peer_port":54109,"peer_addr":"127.0.0.1","route":"{rabbit@sunnyside,'rabbitmqcli-932-rabbit@sunnyside'}","timestamp_in_ms":1724083338452}
{"event":"node.node.stats","send_bytes":98043,"recv_bytes":713,"sock_port":25672,"sock_addr":"127.0.0.1","peer_port":54109,"peer_addr":"127.0.0.1","route":"{rabbit@sunnyside,'rabbitmqcli-932-rabbit@sunnyside'}","timestamp_in_ms":1724083343539}
{"event":"node.node.stats","send_bytes":98305,"recv_bytes":713,"sock_port":25672,"sock_addr":"127.0.0.1","peer_port":54109,"peer_addr":"127.0.0.1","route":"{rabbit@sunnyside,'rabbitmqcli-932-rabbit@sunnyside'}","timestamp_in_ms":1724083348634}
{"event":"node.node.stats","send_bytes":98567,"recv_bytes":717,"sock_port":25672,"sock_addr":"127.0.0.1","peer_port":54109,"peer_addr":"127.0.0.1","route":"{rabbit@sunnyside,'rabbitmqcli-932-rabbit@sunnyside'}","timestamp_in_ms":1724083353723}
{"event":"node.node.stats","send_bytes":98829,"recv_bytes":721,"sock_port":25672,"sock_addr":"127.0.0.1","peer_port":54109,"peer_addr":"127.0.0.1","route":"{rabbit@sunnyside,'rabbitmqcli-932-rabbit@sunnyside'}","timestamp_in_ms":1724083358775}
{"event":"node.node.stats","send_bytes":99091,"recv_bytes":721,"sock_port":25672,"sock_addr":"127.0.0.1","peer_port":54109,"peer_addr":"127.0.0.1","route":"{rabbit@sunnyside,'rabbitmqcli-932-rabbit@sunnyside'}","timestamp_in_ms":1724083363867}
```

@eliyosef578 without clear evidence (an executable way reproduce) and the lack of other similar reports, I can only conclude that most likely your new environment

1. Does not enable the plugin properly
2. Does use bind the queue correctly (does not specify a correct routing key)

**Reply by eliyosef578** [reply]
Hi @michaelklishin thanks for your response.
i try to running the rabbit with docker and i add 3 files to volume : [definitions.json](https://github.com/user-attachments/files/16706375/definitions.json) ,
[enabled_plugins.txt](https://github.com/user-attachments/files/16706388/enabled_plugins.txt),
[rabbitmq.conf.txt](https://github.com/user-attachments/files/16706433/rabbitmq.conf.txt)

the docker run command is:
docker run -p 5672:5672 -p 15672:15672 -v ${PWD}/rabbitmq.conf:/etc/rabbitmq/rabbitmq.conf -v ${PWD}/definitions.json:/etc/rabbitmq/definitions.json v ${PWD}/enabled_plugins:/etc/rabbitmq/enabled_plugins rabbitmq:3.13.6-management-alpine

the event exchange plugin its enabled and the queue and the binding are created after the init, but the events do not reach to the queue.

if i try to runnig the docker without enabled_plugins file And I do the enable manually the events reach to the queue.

**Reply by michaelklishin** [reply]
@eliyosef578 I have demonstrated that the exchange works as expected and explained what the two most typical problems are.

Take a look at your definitions file. You have bindings that do not have any routing key, which is critically important for topic exchanges such as `amq.rabbitmq.events`. Here are what bindings with a routing key specified look like:

``` json
  "bindings": [
    {
      "source": "amq.rabbitmq.event",
      "vhost": "/",
      "destination": "qq.1",
      "destination_type": "queue",
      "routing_key": "#",
      "arguments": {}
    }
```

and an extract from your `definitions.json`:

``` json
  "bindings": [
    {
      "source": "amq.rabbitmq.event",
      "vhost": "/",
      "destination": "events",
      "destination_type": "binding.created",
      "arguments": {}
    },
    {
      "source": "amq.rabbitmq.event",
      "vhost": "/",
      "destination": "events",
      "destination_type": "exchange.created",
      "arguments": {}
    },
    {
      "source": "amq.rabbitmq.event",
      "vhost": "/",
      "destination": "events",
      "destination_type": "binding.deleted",
      "arguments": {}
    }
  ]
```

Your bindings have an invalid destination type, the supported values are `"queue"` and `"exchange"` (I don't remember if streams use `"queue"` or `"stream"` but certainly nothing uses `"binding.deleted"` which is an event type, not a binding destination type).

My best guess is that you have hand-crafted these definitions. They cannot possibly work (or even pass validation before being imported) regardless of how you enable the plugin.

**Reply by eliyosef578** [reply]
@michaelklishin You are right, I attached the wrong definitions file.
I tried defining things manually and exporting to the definitions file, and then I imported it into the rabbit during the upload stages.
But still the events do not reach the event queue that is produced.
Is it possible that the order in which the operations are carried out in the ascension of the rabbit is important? I see that the import is performed before the plugins upload process

**Reply by eliyosef578** [reply]
@michaelklishin i try to debug with 2 kind of definitions files :
[def1.json](https://github.com/user-attachments/files/16739662/def1.json)
[def.json](https://github.com/user-attachments/files/16739663/def.json)
if i try to load the def.json with only exchange.created bindings, the events reached to the events queue.
if i try to load the def1.json with all the 4 bindings the events are not reached to the events queue.
this is the rabbit log :
[rabbit.log](https://github.com/user-attachments/files/16739667/rabbit.log)

**Reply by michaelklishin** [reply]
Definitions are imported after plugins are enabled.

You can always [import definitions after cluster formation](https://www.rabbitmq.com/docs/definitions) using CLI tools. 

With one node it can be a matter of using `rabbitmqctl await_startup` plus waiting for 10s or something like that.
